src/rules/callableTypesRule.ts

```diff
@@ -29,76 +30,68 @@ export class Rule extends Lint.Rules.AbstractRule {
-        return `Interface has only a call signature — use \`type ${name} = ${sigSuggestion}\` instead.`;
-    }
-
-    public static failureStringForTypeLiteral(sigSuggestion: string): string {
-        return `Type literal has only a call signature — use \`${sigSuggestion}\` instead.`;
+    public static FAILURE_STRING_FACTORY(type: string, sigSuggestion: string) {
+        return `${type} has only a call signature — use \`${sigSuggestion}\` instead.`;
     }
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithWalker(new Walker(sourceFile, this.getOptions()));
+        return this.applyWithFunction(sourceFile, walk);
     }
 }
 
-class Walker extends Lint.RuleWalker {
-    public visitInterfaceDeclaration(node: ts.InterfaceDeclaration) {
-        if (noSupertype(node.heritageClauses)) {
-            this.check(node);
-        }
-        super.visitInterfaceDeclaration(node);
-    }
-
-    public visitTypeLiteral(node: ts.TypeLiteralNode) {
-        this.check(node);
-        super.visitTypeLiteral(node);
-    }
-
-    private check(node: ts.InterfaceDeclaration | ts.TypeLiteralNode) {
-        if (node.members.length === 1 && node.members[0].kind === ts.SyntaxKind.CallSignature) {
-            const call = node.members[0] as ts.CallSignatureDeclaration;
-            if (!call.type) {
-                // Bad parse
-                return;
-            }
-
-            const suggestion = renderSuggestion(call);
-            if (node.kind === ts.SyntaxKind.InterfaceDeclaration) {
-                this.addFailureAtNode(node.name, Rule.failureStringForInterface(node.name.getText(), suggestion));
-            } else {
-                this.addFailureAtNode(call, Rule.failureStringForTypeLiteral(suggestion));
+function walk(ctx: Lint.WalkContext<void>) {
+    return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if ((isInterfaceDeclaration(node) && noSupertype(node)
+             || isTypeLiteralNode(node))
+            && node.members.length === 1) {
+            const member = node.members[0];
+            if (isCallSignatureDeclaration(member) &&
+                // avoid bad parse
+                member.type !== undefined) {
+                const suggestion = renderSuggestion(member, node, ctx.sourceFile);
+                ctx.addFailureAtNode(
+                    member,
+                    Rule.FAILURE_STRING_FACTORY(node.kind === ts.SyntaxKind.TypeLiteral ? "Type literal" : "Interface", suggestion),
+                    Lint.Replacement.replaceNode(node, suggestion),
+                );
             }
         }
-    }
+        return ts.forEachChild(node, cb);
+    });
 }
 
 /** True if there is no supertype or if the supertype is `Function`. */
-function noSupertype(heritageClauses: ts.NodeArray<ts.HeritageClause> | undefined): boolean {
-    if (!heritageClauses) {
+function noSupertype(node: ts.InterfaceDeclaration): boolean {
```

Comment:
it makes the signature shorter and the intention of the function clearer. Of course it could be destructured directly in the parameter, but I doubt this would make a difference